### ISSUE TITLE
top_retain results in TypeError: Cannot read property '_top' of undefined

### ISSUE THREAD

**Issue by opichals** [terser-js/terser#183]
**Bug report**

**Version (complete output of `terser -V`)**

3.10.12

**Complete CLI command or `minify()` options used**

... top_retain ...

**`terser` input**

**`terser` output or error**

```
TypeError: Cannot read property '_top' of undefined
    at Zn (/Users/standa/Prg/rollup-plugin-espruino-modules/node_modules/terser/dist/bundle.js:44:164713)
    at /Users/standa/Prg/rollup-plugin-espruino-modules/node_modules/terser/dist/bundle.js:44:184171
    at AST_Call.optimize (/Users/standa/Prg/rollup-plugin-espruino-modules/node_modules/terser/dist/bundle.js:44:121279)
    at Ft.before (/Users/standa/Prg/rollup-plugin-espruino-modules/node_modules/terser/dist/bundle.js:44:121036)
    at AST_Call.transform (/Users/standa/Prg/rollup-plugin-espruino-modules/node_modules/terser/dist/bundle.js:44:78398)
    at /Users/standa/Prg/rollup-plugin-espruino-modules/node_modules/terser/dist/bundle.js:44:78534
```

**Expected result**

**Comment by kzc** [comment]
As per the issue template, the bug report will be closed if terser input and options are not provided to reproduce the problem.

@fabiosantoscode side note: Notice the stack trace is mangled because the bundle source map is not being used by the `minify()` caller. Should consider previous scheme of replacing `-mc -b beautify=false,ascii_only` with `-c defaults=false -b beautify=true,ascii_only`. Alternatively with some effort, a program can be written to unmangle such a stack trace using the bundle source map and the `file:line:col` info per stack frame using the `source-map` API: [`SourceMapConsumer.prototype.originalPositionFor(generatedPosition)`](https://github.com/mozilla/source-map#sourcemapconsumerprototypeoriginalpositionforgeneratedposition).

**Comment by kzc** [comment]
https://www.npmjs.com/package/stack-beautifier

```
$ node_modules/.bin/stack-beautifier node_modules/terser/dist/bundle.js.map -t stack_trace.txt
TypeError: Cannot read property '_top' of undefined
  at _top (./compress.js:4559:18)
  at retain_top_func (./compress.js:4573:16)
  at optimizer (./compress.js:281:22)
  at optimize (./compress.js:264:23)
  at before (./transform.js:61:34)
  at transform (./transform.js:77:24)
```

**Comment by kzc** [comment]
@opichals Can you try this patch and report your findings? You'll have to build `terser` locally. Modifying `node_modules/terser/lib` directly no longer works.

```patch
--- a/lib/compress.js
+++ b/lib/compress.js
@@ -4557,4 +4557,5 @@ merge(Compressor.prototype, {
     function retain_top_func(fn, compressor) {
         return compressor.top_retain
+            && fn
             && fn._top
             && fn instanceof AST_Defun
```
A bug repro test case for the test suite would still be useful if you can provide it.

**Comment by kzc** [comment]
Worth noting that this bug would not occur with default minify options.

**Comment by kzc** [comment]
I was able to come up with a failing test case. Added to #184.

**Comment by fabiosantoscode** [comment]
Should we not minify the output? The performance increase is small, and bug reports are harder to transmit.

**Comment by fabiosantoscode** [comment]
At least not mangling variable and function names would help, I believe.

**Comment by opichals** [comment]
@kzc Thank you! I would not be that quick to isolate from the minified code I had.

Verified, the PR fixes this issue.

**Comment by kzc** [comment]
@fabiosantoscode I don't follow. `top_retain` specifies which top level symbols which shouldn't be optimized away. Anyway, this is just a bug fix. No new functionality.

**Comment by fabiosantoscode** [comment]
@kzc I meant not minifying or not mangling our exported modules. It may help debugging

**Comment by kzc** [comment]
@fabiosantoscode I see what you mean. [`stack-beautifier`] works well enough in my opinion. The names of the exports like `minify` rarely are of significance in stack traces from the bundle. The line numbers are the most important thing to debug issues.

If downstream Terser wrappers would make use of `source-map-support` we wouldn't have this issue. The responsibility to add `source-map-support` is at the application level - as you see in `bin/uglifyjs`. It shouldn't be in the bundle.

**Comment by fabiosantoscode** [comment]
Fix released as v3.11.0

**Comment by fabiosantoscode** [comment]
Well, given there's no performance improvement in mangling, I think we should stop. It's at least a bit better at least for me to read function names when available.

I understand that about source-map-support. We should add a note to README then.
